db/database.go

```diff
@@ -712,13 +716,79 @@ func (context *DatabaseContext) Close(ctx context.Context) {
 	if context.SGReplicateMgr != nil {
 		context.SGReplicateMgr.Stop()
 	}
+
+	waitForBackgroundManagersToStop(ctx, BGTCompletionMaxWait, bgManagers)
+
 	context.Bucket.Close()
 	context.Bucket = nil
 
 	base.RemovePerDbStats(context.Name)
 
 }
 
+// stopBackgroundManagers stops any running BackgroundManager.
+// Returns a list of BackgroundManager it signalled to stop
+func (context *DatabaseContext) stopBackgroundManagers() []*BackgroundManager {
+	bgManagers := make([]*BackgroundManager, 0)
+
+	if context.ResyncManager != nil {
+		if !isBackgroundManagerStopped(context.ResyncManager.GetRunState()) {
+			if err := context.ResyncManager.Stop(); err == nil {
+				bgManagers = append(bgManagers, context.ResyncManager)
+			}
+		}
+	}
+
+	if context.AttachmentCompactionManager != nil {
+		if !isBackgroundManagerStopped(context.AttachmentCompactionManager.GetRunState()) {
+			if err := context.AttachmentCompactionManager.Stop(); err == nil {
+				bgManagers = append(bgManagers, context.AttachmentCompactionManager)
+			}
+		}
+	}
+
+	if context.TombstoneCompactionManager != nil {
+		if !isBackgroundManagerStopped(context.TombstoneCompactionManager.GetRunState()) {
+			if err := context.TombstoneCompactionManager.Stop(); err == nil {
+				bgManagers = append(bgManagers, context.TombstoneCompactionManager)
+			}
+		}
+	}
+
+	return bgManagers
+}
+
+// waitForBackgroundManagersToStop wait for given all BackgroundManagers to stop within given time
+func waitForBackgroundManagersToStop(ctx context.Context, waitTimeMax time.Duration, bgManagers []*BackgroundManager) {
+	timeout := time.NewTicker(waitTimeMax)
+	interval := time.NewTicker(1 * time.Second)
+	defer timeout.Stop()
+	defer interval.Stop()
+
+	for {
+		select {
+		case <-timeout.C:
+			base.WarnfCtx(ctx, "Timeout after %v of waiting for Background Managers to terminate", waitTimeMax)
```

Comment:
I agree too. However, the problem is that there is no way in current implementation to know what BackgroundManager is doing. In order to log which BackgroundManager failed to stop within given time limit, we would have to change `BackgroundManagerProcessI` to add `Name` method so that we can leverage that in logging. 

Another solution could be what I did in https://github.com/couchbase/sync_gateway/pull/5870/commits/49ec67bc0bc7b4e77dc278f89ef96e9b8f6801ab#diff-d1f45819145818652adec732708bb3dc6640a0f966b4287696eeb45ab9c901eaR733 where `stopBackgroundProcess` was taking in `processName` to log this kind of messages (signalled to stop, failed to stop, stopped, etc). 

Personally, I'd prefer to not change `BackgroundManagerProcessI` interface just for logging purpose. Let me know what are your thoughts on this.

Reply:
As discussed, let's add a name to BackgroundManager struct and API for this purpose.

Reply:
addressed